Demo-Swift/Demo-Swift/MHAudioBufferPlayer/MHAudioBufferPlayer.swift

```diff
@@ -0,0 +1,160 @@
+typealias MHAudioBufferPlayerBlock = (_ buffer: AudioQueueBufferRef, _ audioFormat: AudioStreamBasicDescription) -> Void
+let NumberOfAudioDataBuffers = 3
+
+func InterruptionListenerCallback(inUserData: UnsafeRawPointer, interruptionState: UInt32) {
+  let player = Unmanaged<MHAudioBufferPlayer>.fromOpaque(inUserData).takeUnretainedValue()
+  if interruptionState == kAudioSessionBeginInterruption {
+    player.tearDownAudio()
+  } else if interruptionState == kAudioSessionEndInterruption {
+    player.setUpAudio()
+    player.start()
+  }
+}
+
+func PlayCallback(_ inUserData: UnsafeMutableRawPointer!, _ inAudioQueue: AudioQueueRef, _ inBuffer: AudioQueueBufferRef) {
+  let player = Unmanaged<MHAudioBufferPlayer>.fromOpaque(inUserData).takeUnretainedValue()
+  if player.playing && player.block != nil {
+    player.block!(inBuffer, player.audioFormat)
+    AudioQueueEnqueueBuffer(inAudioQueue, inBuffer, 0, nil)
+  }
+}
+
+class MHAudioBufferPlayer {
+  
+  var block: MHAudioBufferPlayerBlock?
+  private(set) var playing: Bool
+  var gain: Float32
+  var audioFormat: AudioStreamBasicDescription
+  
+  var _playQueue: AudioQueueRef?
+  var _playQueueBuffers: [AudioQueueBufferRef?] = .init(repeating: nil, count: NumberOfAudioDataBuffers)
+  var _packetsPerBuffer: UInt32
+  var _bytesPerBuffer: UInt32
+  
+  convenience init(sampleRate: Float64, channels: UInt32, bitsPerChannel: UInt32, secondsPerBuffer: Float64) {
+    self.init(sampleRate: sampleRate, channels: channels, bitsPerChannel: bitsPerChannel, packetsPerBuffer: UInt32(secondsPerBuffer * sampleRate))
+  }
+  
+  init(sampleRate: Float64, channels: UInt32, bitsPerChannel: UInt32, packetsPerBuffer: UInt32) {
+    playing = false
+    _playQueue = nil
+    gain = 1.0
+    
+    audioFormat = AudioStreamBasicDescription()
+    
+    audioFormat.mFormatID = kAudioFormatLinearPCM
+    audioFormat.mSampleRate = sampleRate
+    audioFormat.mChannelsPerFrame = channels
+    audioFormat.mBitsPerChannel = bitsPerChannel
+    audioFormat.mFramesPerPacket = 1 // uncompressed audio
+    audioFormat.mBytesPerFrame = audioFormat.mChannelsPerFrame * audioFormat.mBitsPerChannel / 8
+    audioFormat.mFormatFlags = kLinearPCMFormatFlagIsSignedInteger | kLinearPCMFormatFlagIsPacked
+    
+    _packetsPerBuffer = packetsPerBuffer
+    _bytesPerBuffer = _packetsPerBuffer * audioFormat.mBytesPerPacket
+    
+    setUpAudio()
+  }
+  
+  deinit {
+    tearDownAudio()
+  }
+  
+  func setUpAudio() {
+    if _playQueue == nil {
+      setUpAudioSession()
+      setUpPlayQueue()
+      setUpPlayQueueBuffers()
+    }
+  }
+  
+  func tearDownAudio() {
+    if _playQueue != nil {
+      stop()
+      tearDownPlayQueue()
+      tearDownAudioSession()
+    }
+  }
+  
+  func setUpAudioSession() {
+    
+    // TODO: - Verify that `AudioQueueInitialize` is not necessary.
+    
+    do {
+      try AVAudioSession.sharedInstance().setActive(true)
+    } catch {
+      print("Audio session setup failed: \(error)")
+    }
+  }
```

Comment:
The original Objective-C implementation was the following:

```
- (void)setUpAudioSession
{
  AudioSessionInitialize(
    NULL,
    NULL,
    InterruptionListenerCallback,
    (__bridge void *)self);

  UInt32 sessionCategory = kAudioSessionCategory_MediaPlayback;
    AudioSessionSetProperty(
      kAudioSessionProperty_AudioCategory,
      sizeof(sessionCategory),
      &sessionCategory);

  AudioSessionSetActive(true);
}
```

`AudioSessionInitialize` has been deprecated since iOS 7.0. Other users have tried addressing this problem. According to this mini blog [post](https://github.com/software-mariodiana/AudioBufferPlayer/wiki/Replacing-C-functions-deprecated-in-iOS-7), they suggest using `AVAudioSession` to listen for interruptions.

However, the described procedure (the use of `AVAudioSession.sharedInstance().delegate`) has also been deprecated. 

I was unable to find a suitable replacement, so I could use some pointers here.